api/src/pcapi/alembic/versions/20220106T152851_b85084de1d58_change_database_check_booking_trigger.py

```diff
@@ -0,0 +1,132 @@
+"""change_database_check_booking_trigger
+"""
+from alembic import op
+
+
+# revision identifiers, used by Alembic.
+revision = "b85084de1d58"
+down_revision = "525b100612c2"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    op.execute(
+        """
+          CREATE OR REPLACE FUNCTION check_booking()
+    RETURNS TRIGGER AS $$
+    DECLARE
+        lastStockUpdate date := (SELECT "dateModified" FROM stock WHERE id=NEW."stockId");
+        deposit_id bigint := (SELECT individual_booking."depositId" FROM booking LEFT JOIN individual_booking ON individual_booking.id = booking."individualBookingId" WHERE booking.id=NEW.id);
+    BEGIN
+    IF EXISTS (SELECT "quantity" FROM stock WHERE id=NEW."stockId" AND "quantity" IS NOT NULL)
+        AND (
+            (SELECT "quantity" FROM stock WHERE id=NEW."stockId")
+            <
+            (SELECT SUM(quantity) FROM booking WHERE "stockId"=NEW."stockId" AND status != 'CANCELLED')
+            )
+        THEN RAISE EXCEPTION 'tooManyBookings'
+                    USING HINT = 'Number of bookings cannot exceed "stock.quantity"';
+    END IF;
+
+    IF (
+        (NEW."educationalBookingId" IS NULL AND OLD."educationalBookingId" IS NULL)
+        AND (NEW."individualBookingId" IS NOT NULL OR OLD."individualBookingId" IS NOT NULL)
+        AND (
+        -- If this is a new booking, we probably want to check the wallet.
+        OLD IS NULL
+        -- If we're updating an existing booking...
+        OR (
+            -- Check the wallet if we are changing the quantity or the amount
+            -- The backend should never do that, but let's be defensive.
+            (NEW."quantity" != OLD."quantity" OR NEW."amount" != OLD."amount")
+            -- If amount and quantity are unchanged, we want to check the wallet
+            -- only if we are UNcancelling a booking. (Users with no credits left
+            -- should be able to cancel their booking. Also, their booking can
+            -- be marked as used or not used.)
+            OR (NEW.status != OLD.status AND NEW.status != 'CANCELLED')
```

Comment:
Je me demande si la logique est respectée là, car on dit au dessus 
```
      -- If amount and quantity are unchanged, we want to check the wallet
      -- only if we are UNcancelling a booking.
```

et dans la version avant ta modif on fait:
`OR (NEW."isCancelled" != OLD."isCancelled" AND NOT NEW."isCancelled")`

Ce qui n'est pas exactement identique en terme logique je pense.
Imaginons un `NEW.status` à `USED`, un ancien à `CONFIRMED`, alors ce or renverrait `true`, là où l'ancien renverrait false car les deux seraient à `isCancelled = false`

Je ne sais pas si ça peut avoir des répercussions sur certains use cases, mais peut-être qu'il faut coller à l'ancien comportement?

Reply:
Corrigé